File: RpgSystem/CAttribute.cpp
/** 
 *  AKARRA SOURCE CODE (c) Jens Bergensten
 *  See EULA.txt for information about this file.
 *
 *  File description (may not be accurate):
 *    Attributes are the basic stats of an entity, but are usually not
 *    used in calculations. Instead they are derived to abilities.
 **/

#include "CAttribute.h"

namespace RpgSystem
{

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::CAttribute() (constructor)
//
// Description:	Default constructor. Clears all member data.
// Parameters:	None.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
CAttribute::CAttribute()
:
	m_base(0),
	m_bonus(0),
	m_percentbonus(0)
{
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::~CAttribute() (destructor)
//
// Description:	Default destructor. Removes acquired memory.
// Parameters:	None.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
CAttribute::~CAttribute()
{
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::setBase()
//
// Description:	Sets the base attribute value.
// Parameters:	float base:	The requested value.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CAttribute::setBase(float base)
{
	m_base = base;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::resetBonus()
//
// Description:	Resets the bonus to 0, so bonus-adding can be started.
// Parameters:	None.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CAttribute::resetBonus()
{
	m_bonus = 0;
	m_percentbonus = 0;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::addBonus()
//
// Description:	Adds a value to the bonus.
// Parameters:	float bonus:	The the amount to increase bonus with.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CAttribute::addBonus(float bonus)
{
	m_bonus += bonus;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::addPercentBonus()
//
// Description:	Adds a value to the percent bonus.
// Parameters:	float bonus:	The the amount to increase bonus with. Note that this is
//								a percentile bonus, so a value of 0.2f would increase
//								the bonus by 20%.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CAttribute::addPercentBonus(float bonus)
{
	m_percentbonus += bonus;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::getBase()
//
// Description:	Used to get the base value.
// Parameters:	None.
// Returns:		The base value (float).
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
float CAttribute::getBase()
{
	return m_base;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::getBonus()
//
// Description:	Used to get the bonus value.
// Parameters:	None.
// Returns:		The bonus value (float).
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
float CAttribute::getBonus()
{
	return m_bonus;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::getPercentBonus()
//
// Description:	Used to get the percentile bonus value.
// Parameters:	None.
// Returns:		The bonus value (float).
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
float CAttribute::getPercentBonus()
{
	return m_percentbonus;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CAttribute::getValue()
//
// Description:	Used to get the total attribute value.
// Parameters:	None.
// Returns:		The total value (float).
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
float CAttribute::getValue()
{
	float value = (m_base + m_bonus) * (m_percentbonus + 1.0f);
	return (value < 0) ? 0.0f : value;
}

bool CAttribute::read(FileSystem::CFileStream& file, int version)
{
	if (!file.isOk())
	{
		return false;
	}

	file.read(&m_base, sizeof(float), 1);
	file.read(&m_bonus, sizeof(float), 1);
	file.read(&m_percentbonus, sizeof(float), 1);

	return true;
}

bool CAttribute::write(FileSystem::CFileStream& file, int version)
{
	if (!file.isOk())
	{
		return false;
	}

	file.write(&m_base, sizeof(float), 1);
	file.write(&m_bonus, sizeof(float), 1);
	file.write(&m_percentbonus, sizeof(float), 1);

	return true;
}

// ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** *****
//  ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ****
// ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** *****

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CVarAttribute::CVarAttribute() (constructor)
//
// Description:	Default constructor. Clears all member data.
// Parameters:	None.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
CVarAttribute::CVarAttribute()
:
	CAttribute(),
	m_current(0)
{
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CVarAttribute::~CVarAttribute() (destructor)
//
// Description:	Default destructor. Removes acquired memory.
// Parameters:	None.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
CVarAttribute::~CVarAttribute()
{
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CVarAttribute::setBase()
//
// Description:	Sets the base max value. This only affects the current value if the new
//				max is less than the current value.
// Parameters:	float value:	The new max value.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CVarAttribute::setBase(float value)
{
	m_base = value;

	if (m_current > getValue())
	{
		m_current = getValue();
	}
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CVarAttribute::resetMaxBonus()
//
// Description:	Resets the bonus to 0. Reduces the current value because we do not want
//				exploits.
// Parameters:	None.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CVarAttribute::resetBonus()
{
	// get delta value
	float old = getValue();

	m_bonus = 0;
	m_percentbonus = 0;

	m_current -= getValue() - old;

	if (m_current > getValue())
	{
		m_current = getValue();
	}
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CVarAttribute::addBonus()
//
// Description:	Increases the bonus of the maximum score. Also increases the current
//				value (and makes sure it isn't more than the new maximum).
// Parameters:	float amount:	The amount to be added.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CVarAttribute::addBonus(float amount)
{
	// get delta value
	float old = getValue();

	m_bonus += amount;
	m_current += getValue() - old;

	if (m_current > getValue())
	{
		m_current = getValue();
	}
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CVarAttribute::addPercentBonus()
//
// Description:	Increases the percentile bonus of the maximum score. Also increases the
//				current value (and makes sure it isn't more than the new maximum).
// Parameters:	float amount:	The amount to be added.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CVarAttribute::addPercentBonus(float amount)
{
	// get delta value
	float old = getValue();

	m_percentbonus += amount;
	m_current += getValue() - old;

	if (m_current > getValue())
	{
		m_current = getValue();
	}
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CVarAttribute::getCurrent()
//
// Description:	Returns the current value.
// Parameters:	None.
// Returns:		The current value.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
float CVarAttribute::getCurrent()
{
	return m_current;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// Description:	
// Parameters:	
// Returns:		
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CVarAttribute::setCurrent(float value)
{
	m_current = (value <= getValue()) ? value : getValue();
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CVarAttribute::addToCurrent()
//
// Description:	Adds a value to the current.
// Parameters:	float amount:	The value to be added (use negative values for, say,
//								damage).
// Returns:		The new current value.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
float CVarAttribute::addToCurrent(float amount)
{
	m_current += amount;

	if (m_current > getValue())
	{
		m_current = getValue();
	}

	return m_current;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// CVarAttribute::restore()
//
// Description:	Restores the current value back to max.
// Parameters:	None.
// Returns:		Nothing.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CVarAttribute::restore()
{
	m_current = getValue();
}

bool CVarAttribute::read(FileSystem::CFileStream& file, int version)
{
	if (!file.isOk())
	{
		return false;
	}

	file.read(&m_current, sizeof(float), 1);

	return CAttribute::read(file, version);
}

bool CVarAttribute::write(FileSystem::CFileStream& file, int version)
{
	if (!file.isOk())
	{
		return false;
	}

	file.write(&m_current, sizeof(float), 1);

	return CAttribute::write(file, version);
}


// ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** *****
//  ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ****
// ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** *****

const char* ATTRIBUTE_NAME[NUM_ATTRIBUTES] = 
{
	"Strength",
	"Constitution",
	"Agility",
	"Intelligence",
	"Faith"
};

const char* VARATTRIBUTE_NAME[NUM_VARATTRIBUTES] = 
{
	"Health",
	"Mana",
	"Stamina"
};

// ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** *****
//  ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ****
// ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** ***** *****

char* getAttributeName(int attribute)
{
	return (char*)ATTRIBUTE_NAME[attribute];
}

char* getVarAttributeName(int attribute)
{
	return (char*)VARATTRIBUTE_NAME[attribute];
}

} // end of RpgSystem namespace

File: Src/RPG-System/Attributes/CAttribute.cs
using System;
using System.IO;

namespace RpgSystem.Attributes
{
    /// <summary>
    /// Base attribute class that forms the foundation of the character attribute system.
    /// Attributes have a base value, bonus value, and percentage bonus.
    /// </summary>
    public class CAttribute
    {
        // Fields
        protected float _baseValue;
        protected float _bonus;
        protected float _percentBonus;

        /// <summary>
        /// Creates a new attribute with default values (0)
        /// </summary>
        public CAttribute()
        {
            _baseValue = 0;
            _bonus = 0;
            _percentBonus = 0;
        }

        /// <summary>
        /// Creates a new attribute with a specified base value
        /// </summary>
        public CAttribute(float baseValue)
        {
            _baseValue = baseValue;
            _bonus = 0;
            _percentBonus = 0;
        }

        /// <summary>
        /// Sets the base value of the attribute
        /// </summary>
        public virtual void SetBase(float value)
        {
            _baseValue = value;
        }

        /// <summary>
        /// Resets all bonuses to zero
        /// </summary>
        public virtual void ResetBonus()
        {
            _bonus = 0;
            _percentBonus = 0;
        }

        /// <summary>
        /// Adds a flat bonus to the attribute
        /// </summary>
        public virtual void AddBonus(float value)
        {
            _bonus += value;
        }

        /// <summary>
        /// Adds a percentage bonus to the attribute (0.1 = 10%)
        /// </summary>
        public virtual void AddPercentBonus(float value)
        {
            _percentBonus += value;
        }

        /// <summary>
        /// Gets the base value of the attribute
        /// </summary>
        public virtual float GetBase()
        {
            return _baseValue;
        }

        /// <summary>
        /// Gets the flat bonus value of the attribute
        /// </summary>
        public virtual float GetBonus()
        {
            return _bonus;
        }

        /// <summary>
        /// Gets the percentage bonus of the attribute
        /// </summary>
        public virtual float GetPercentBonus()
        {
            return _percentBonus;
        }

        /// <summary>
        /// Gets the total value of the attribute (base + bonus) * (1 + percentBonus)
        /// </summary>
        public virtual float GetValue()
        {
            float value = (_baseValue + _bonus) * (1.0f + _percentBonus);
            return value < 0 ? 0 : value;
        }

        /// <summary>
        /// Reads attribute data from a binary stream
        /// </summary>
        public virtual bool Read(BinaryReader reader, int version)
        {
            try
            {
                _baseValue = reader.ReadSingle();
                _bonus = reader.ReadSingle();
                _percentBonus = reader.ReadSingle();
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }

        /// <summary>
        /// Writes attribute data to a binary stream
        /// </summary>
        public virtual bool Write(BinaryWriter writer, int version)
        {
            try
            {
                writer.Write(_baseValue);
                writer.Write(_bonus);
                writer.Write(_percentBonus);
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }
    }

    /// <summary>
    /// Extended attribute class that has a current value that can be different from its maximum.
    /// Used for health, mana, stamina, etc.
    /// </summary>
    public class CVariableAttribute : CAttribute
    {
        // Fields
        protected float _current;

        /// <summary>
        /// Creates a new variable attribute with default values
        /// </summary>
        public CVariableAttribute() : base()
        {
            _current = 0;
        }

        /// <summary>
        /// Creates a new variable attribute with a specified base value
        /// </summary>
        public CVariableAttribute(float baseValue) : base(baseValue)
        {
            _current = GetValue();
        }

        /// <summary>
        /// Sets the base value and updates current value if needed
        /// </summary>
        public override void SetBase(float value)
        {
            float oldMax = GetValue();
            base.SetBase(value);
            float newMax = GetValue();

            // If max value increased, increase current value by the same amount
            if (newMax > oldMax)
            {
                _current += (newMax - oldMax);
            }
            // If max value decreased and current is now greater than max, cap it
            else if (_current > newMax)
            {
                _current = newMax;
            }
        }

        /// <summary>
        /// Resets bonuses and adjusts current value if necessary
        /// </summary>
        public override void ResetBonus()
        {
            float oldMax = GetValue();
            base.ResetBonus();
            float newMax = GetValue();

            // Adjust current value proportionally to change in max value
            if (oldMax > 0)
            {
                _current = (_current / oldMax) * newMax;
            }
            else
            {
                _current = newMax;
            }
        }

        /// <summary>
        /// Adds a bonus and adjusts current value proportionally
        /// </summary>
        public override void AddBonus(float value)
        {
            float oldMax = GetValue();
            base.AddBonus(value);
            float newMax = GetValue();
            float diff = newMax - oldMax;

            // If max value increased, increase current by the same amount
            if (diff > 0)
            {
                _current += diff;
            }
            // If max value decreased and current is now greater than max, cap it
            else if (_current > newMax)
            {
                _current = newMax;
            }
        }

        /// <summary>
        /// Adds a percentage bonus and adjusts current value proportionally
        /// </summary>
        public override void AddPercentBonus(float value)
        {
            float oldMax = GetValue();
            base.AddPercentBonus(value);
            float newMax = GetValue();
            float diff = newMax - oldMax;

            // If max value increased, increase current by the same amount
            if (diff > 0)
            {
                _current += diff;
            }
            // If max value decreased and current is now greater than max, cap it
            else if (_current > newMax)
            {
                _current = newMax;
            }
        }

        /// <summary>
        /// Gets the current value
        /// </summary>
        public virtual float GetCurrent()
        {
            return _current;
        }

        /// <summary>
        /// Sets the current value (capped at maximum)
        /// </summary>
        public virtual void SetCurrent(float value)
        {
            float max = GetValue();
            _current = value > max ? max : value;
        }

        /// <summary>
        /// Adds to the current value (capped at maximum)
        /// </summary>
        public virtual float AddToCurrent(float value)
        {
            _current += value;
            float max = GetValue();

            if (_current > max)
            {
                _current = max;
            }
            else if (_current < 0)
            {
                _current = 0;
            }

            return _current;
        }

        /// <summary>
        /// Restores the current value to maximum
        /// </summary>
        public virtual void Restore()
        {
            _current = GetValue();
        }

        /// <summary>
        /// Gets the percentage of current to maximum value (0.0 to 1.0)
        /// </summary>
        public virtual float GetPercentage()
        {
            float max = GetValue();
            return max > 0 ? _current / max : 0;
        }

        /// <summary>
        /// Reads variable attribute data from a binary stream
        /// </summary>
        public override bool Read(BinaryReader reader, int version)
        {
            try
            {
                _current = reader.ReadSingle();
                return base.Read(reader, version);
            }
            catch (Exception)
            {
                return false;
            }
        }

        /// <summary>
        /// Writes variable attribute data to a binary stream
        /// </summary>
        public override bool Write(BinaryWriter writer, int version)
        {
            try
            {
                writer.Write(_current);
                return base.Write(writer, version);
            }
            catch (Exception)
            {
                return false;
            }
        }
    }
}